ops/relation.py

```diff
@@ -0,0 +1,548 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import semantic_version as semver
+
+from ops.charm import CharmEvents
+from ops.framework import (
+    Object, EventSource, EventBase, StoredState
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Provider(Object):
+    """Manages relations of resource providers
+
+    A :class:`Provider` object manages relations of a resource provider
+    charms, with charms that consume those resources. When a consumer
+    charm joins the relation a provider object informs it of the type
+    and version of the resource provided. Providers may also be
+    instrumented to inform consumer charms of any relavent
```

Comment:
```suggestion
    instrumented to inform consumer charms of any relevant
```

Overall there are quite a few typos / spelling mistakes, For now I want to consider the overall design vs specific details, but it is something that needs a textual review for clarity.

Reply:
Agreed. Will run spell check and fix all. But please do comment on any semantic ambiguity or awkward sentence construction.

Reply:
Complete file was spellchecked and corrected.